src/dmd/root/array.d

```diff
@@ -803,10 +803,262 @@ if (is(ReturnType!(typeof((T t) => callable(t))) == int))
+        static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
+        {
+            if (range1.length != range2.length)
+                return false;
+        }
+
+        else
+        {
+            for (; !range1.empty; range1.popFront(), range2.popFront())
+            {
+                if (range2.empty)
+                    return false;
+
+                if (range1.front != range2.front)
+                    return false;
+            }
+        }
+
+        return true;
+    }
+}
+
+///
+pure nothrow @nogc @safe unittest
+{
+    enum a = [ 1, 2, 4, 3 ].staticArray;
+    static assert(!equal(a[], a[1..$]));
+    static assert(equal(a[], a[]));
+
+    // different types
+    enum b = [ 1.0, 2, 4, 3].staticArray;
+    static assert(!equal(a[], b[1..$]));
+    static assert(equal(a[], b[]));
+}
+
+/**
+ * Lazily filters the given range based on the given predicate.
+ *
+ * Returns: a range containing only elements for which the predicate returns
+ *  `true`
+ */
+auto filter(alias predicate, Range)(Range range)
+{
+    return Filter!(predicate, Range)(range);
+}
+
+///
+pure nothrow @safe @nogc unittest
+{
+    enum a = [1, 2, 3, 4].staticArray;
+    enum result = a[].filter!(e => e > 2);
+
+    enum expected = [3, 4].staticArray;
+    static assert(result.equal(expected[]));
+}
+
+private struct Filter(alias predicate, Range)
+{
+    private Range range;
+    private bool primed;
+
+    private void prime()
+    {
+        if (primed)
+            return;
+
+        while (!range.empty && !predicate(range.front))
+            range.popFront();
+
+        primed = true;
+    }
+
+    @property bool empty()
+    {
+        prime();
+        return range.empty;
+    }
+
+    @property auto front()
+    {
+        prime();
+        return range.front;
+    }
+
+    void popFront()
+    {
+        prime();
+
+        do
+        {
+            range.popFront();
+        } while (!range.empty && !predicate(range.front));
+    }
+}
+
+/**
+ * Lazily iterates the given range and calls the given callable for each element.
+ *
+ * Returns: a range containing the result of each call to `callable`
+ */
+auto map(alias callable, Range)(Range range)
+if (isInputRange!(Unqual!Range))
+{
+    return Map!(callable, Range)(range);
+}
+
+///
+pure nothrow @safe @nogc unittest
+{
+    enum a = [1, 2, 3, 4].staticArray;
+    enum expected = [2, 4, 6, 8].staticArray;
+
+    enum result = a[].map!(e => e * 2);
+    static assert(result.equal(expected[]));
+}
+
+private struct Map(alias callable, Range)
+{
+    private Range range;
+
+    @property auto front()
+    {
+        return callable(range.front);
+    }
+
+    @property bool empty()
+    {
+        return range.empty;
+    }
+
+    void popFront()
+    {
+        range.popFront();
+    }
+
+    static if (hasLength!(Unqual!Range))
+    {
+        @property auto length()
+        {
+            return range.length;
+        }
+
+        alias opDollar = length;
+    }
```

Comment:
I would add `opSlice` overloads, s.t. the ranges can be saved/copied (I see that you have already used this for built-in arrays)

Reply:
That looks like it's going to be a bit more complicated. As in quite a bit of extra code need to be added.

Reply:
Fair enough, we can always add it when needed.